Adding Availability model to system.

kgf opened the pull request:
This is a big commit for a complex area. Certainly not finished in this
area, but major functionality is here, such as tracking vacations
across events. Will also need additional tests to test when an
availability and an event overlap.

kgf committed:
Adding Availability model to system.

This is a big commit for a complex area. Certainly not finished in this
area, but major functionality is here, such as tracking vacations
acrooss events. Will also need additional tests to test when an
availability and an event overlap.

david135 commented on app/assets/javascripts/application.js:
```diff
@@ -17,4 +17,7 @@
 //= require jquery-ui-timepicker-addon
+//= require datetimepicker
+//= require jquery-ui
+//= require bootstrap
```
According to [Bootstrap](https://github.com/twbs/bootstrap-sass#a-ruby-on-rails) (toward the bottom of that section):

> bootstrap-sprockets and bootstrap should not both be included in application.js.
> 
> bootstrap-sprockets provides individual Bootstrap Javascript files (alert.js or dropdown.js, for example), while bootstrap provides a concatenated file containing all Bootstrap Javascripts.

They [link](https://github.com/twbs/bootstrap-sass/issues/829#issuecomment-75153827) to an issue thread explaining it.

kgf replied:
Awesome catch, thanks. I'll remove that line.

david135 commented on app/views/availabilities/show.html.erb:
```diff
@@ -0,0 +1,32 @@
+</p>
+
+<% if @availability.event.present? %>
+  <p><b>Event:</b><%= link_to @availability.event.title, @availability.event %></p>
```
We could use a space here between </b> and <%= link_to

david135 commented on app/views/availabilities/show.html.erb:
```diff
@@ -0,0 +1,32 @@
+
+<p>
+  <strong>End time:</strong>
+  <%= @availability.end_time %>
```
Is there a way to get the javascript to automatically bring the end time to no earlier than the start time?

kgf replied:
That is a good idea, and may be handled partially by the calendar library. I will create an issue, as I don't want to hold up this pr for that.

kgf replied:
Issue created #140

david135 commented on app/views/people/show.html.erb:
```diff
@@ -2,6 +2,7 @@
   <%= link_to 'Return to Listing', :back, class: 'btn btn-primary', style: 'width: 100%' %><br/><br/>
   <%= link_to 'Edit Person', edit_person_path(@person), class: 'btn btn-primary', style: 'width: 100%' if can? :edit, @person %><br/><br/>
   <%= link_to 'New Certification', new_person_cert_path(@person), class: 'btn btn-primary', style: 'width: 100%' if can? :create, Cert %><br/><br/>
+  <%= link_to 'New Availability', new_person_availability_path(@person), :class => 'btn btn-primary', style: 'width: 100%' if can? :create, Availability  %>
```
'New Availability' or 'Create Availability' as in events/show.html.erb?

kgf replied:
We should move towards new, it's shorter

kgf committed:
Removed extra call to bootstrap JS

kgf commented on app/controllers/events_controller.rb:
```diff
@@ -28,6 +28,9 @@ def index
   def show
     @event = Event.find(params[:id])
     @page_title = @event.title
+    @eligible = Person.all # In the future, this will need to honor department
```
I'm going to move this into the class

kgf commented on app/models/event.rb:
```diff
@@ -2,53 +2,83 @@ class Event < ActiveRecord::Base
     self.timecards.sum('actual_duration')
   end
-  def unknown_people 
+  def unknown_people
```
This is replaced by unresonsive people above

kgf commented on db/migrate/20160901210947_create_availabilities.rb:
```diff
@@ -0,0 +1,14 @@
+      t.timestamps null: false
+    end
+    add_reference :availabilities, :person, index: true
+    add_reference :availabilities, :event, index: true
```
I'm going to remove this, not certain it is needed.

david135 commented on app/models/event.rb:
```diff
@@ -2,53 +2,83 @@ class Event < ActiveRecord::Base
-    end
-  end
+
   def end_date_cannot_be_before_start
```
Could this be changed to `end_time_cannot_be_before_start` (here and in the `validate` line)?  I'd find it more readable to have this lower in the file, among the other methods, probably in the `private` section.

kgf replied:
For some reason, I was thinking it had to be defined before it could be called. Looking [at Ruby Guides](http://guides.rubyonrails.org/active_record_validations.html#custom-methods) shows that is not the case.

kgf committed:
Listing availabilities on the person show and

corrections in data model. It removes the link from availabilities to
events, which can be derived from time spans.

kgf committed:
Removed association from events to availabilities

And made the Availability factory better

kgf committed:
Added scema.rb to PR

kgf committed:
Further removal of events from availabilities.
